server/interceptor/interceptor.go

```diff
@@ -107,24 +109,24 @@ func getRPCInfo(req interface{}, fullMethod, quotaUser string) (*rpcInfo, error)
 		return nil, status.Errorf(codes.Internal, "cannot retrieve treeID from request: %T", req)
 	}
 
-	serviceName, methodName, err := parseFullMethod(fullMethod)
-	if err != nil {
-		return nil, err
-	}
-
 	var treeType trillian.TreeType
-	switch serviceName {
-	case "trillian.TrillianAdmin":
+	switch {
+	case isAdminRPC(req):
 		treeType = trillian.TreeType_UNKNOWN_TREE_TYPE // unrestricted
-	case "trillian.TrillianLog":
+	case isLogRPC(req):
 		treeType = trillian.TreeType_LOG
-	case "trillian.TrillianMap":
+	case isMapRPC(req):
 		treeType = trillian.TreeType_MAP
 	default:
-		return nil, status.Errorf(codes.Internal, "cannot determine treeType for request: %T", req)
+		return nil, status.Errorf(codes.Internal, "unmapped request type: %T", req)
 	}
 
-	readonly := strings.HasPrefix(methodName, "Get") || strings.HasPrefix(methodName, "List")
+	reqType := fmt.Sprintf("%T", req)
```

Comment:
I'm not keen on this approach -- the magic reflection stuff makes it hard for tools (and humans) to spot the connection between the RPC types and their policing.

Given there is already a set of type switches below, why not extend them to return (e.g.) `(isAdmin bool, readonly bool)`?